repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeServlet.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.powerfilter
+
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.inject.{Inject, Named}
+import javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR
+import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper, ResponseMode}
+
+@Named
+class ReposeServlet @Inject()(router: PowerFilterRouter) extends HttpServlet with StrictLogging {
+
+  override def service(req: HttpServletRequest, resp: HttpServletResponse): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(req)
+    val wrappedResponse = new HttpServletResponseWrapper(resp, ResponseMode.MUTABLE, ResponseMode.PASSTHROUGH)
+
+    try {
+      if (wrappedResponse.getStatus() < SC_INTERNAL_SERVER_ERROR) {
+        router.route(wrappedRequest, wrappedResponse)
+      }
+    } catch {
+      case e: Exception =>
+        logger.error("Failed to route the request to the origin service", e)
```

Comment:
Do we really want to wrap the response? In the chain i mostly checked if the response was committed and did a send error if it wasn't. And always added the log message. I'm not certain which is the better approach here, but i wanted to bring it up for discussion.

Reply:
Fair question. I debated this a bit with myself. I guess the root question is, if an `Exception` is thrown, do we want downstream to be able to write the response, or should the active component be responsible for it?

I chose to put the responsibility on the active component, since if the downstream component threw an exception, processing failed in some way, so I don't trust whatever response was written. In my mind, the exception communicates to the active component that downstream could not handle processing. Admittedly, a competent downstream component would never write a response that should not reach the user. Also, the downstream component may have more context to write a good response. Since the exception is probably some technical problem, though, I'd rather stick whatever context we have in a log, and not send it back to the end user.